lib/secretsscanner/scan/scan_test.go

```diff
@@ -0,0 +1,192 @@
+				got = append(got, key.Key)
+			}
+
+			// Sort the keys by name for comparison.
+			sortPrivateKeys(expect)
+			sortPrivateKeys(got)
+
+			diff := cmp.Diff(expect, got, protocmp.Transform())
+			require.Empty(t, diff, "unexpected keys")
+
+		})
+	}
+}
+
+func sortPrivateKeys(keys []*accessgraphsecretsv1pb.PrivateKey) {
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].Metadata.Name < keys[j].Metadata.Name
+	})
+}
+
+func writeEncryptedKeys(t *testing.T, dir string) []*accessgraphsecretsv1pb.PrivateKey {
+	t.Helper()
+	var expectedKeys []*accessgraphsecretsv1pb.PrivateKey
+	// Write encrypted keys to the directory.
+	for _, key := range cryptosshtestdata.PEMEncryptedKeys {
+		err := os.Mkdir(filepath.Join(dir, key.Name), os.ModePerm)
+		require.NoError(t, err)
+
+		filePath := filepath.Join(dir, key.Name, key.Name)
+		err = os.WriteFile(filePath, key.PEMBytes, os.ModePerm)
+		require.NoError(t, err)
+
+		s, err := ssh.ParsePrivateKeyWithPassphrase(key.PEMBytes, []byte(key.EncryptionKey))
+		require.NoError(t, err)
+
+		if !key.IncludesPublicKey {
+			pubFilePath := filePath + ".pub"
+			authorizedKeyBytes := ssh.MarshalAuthorizedKey(s.PublicKey())
+			require.NoError(t, os.WriteFile(pubFilePath, authorizedKeyBytes, os.ModePerm))
+		}
+
+		fingerprint := ssh.FingerprintSHA256(s.PublicKey())
+
+		mode := accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_DERIVED
+		if !key.IncludesPublicKey {
+			mode = accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_PUB_FILE
+		}
+
+		key, err := accessgraph.NewPrivateKeyWithName(
+			privateKeyNameGen(filePath, deviceID, fingerprint),
+			&accessgraphsecretsv1pb.PrivateKeySpec{
+				PublicKeyFingerprint: fingerprint,
+				DeviceId:             deviceID,
+				PublicKeyMode:        mode,
+			},
+		)
+		require.NoError(t, err)
+
+		expectedKeys = append(expectedKeys, key)
+	}
+
+	return expectedKeys
+}
+
+func writeUnEncryptedKeys(t *testing.T, dir string) []*accessgraphsecretsv1pb.PrivateKey {
+	t.Helper()
+	var expectedKeys []*accessgraphsecretsv1pb.PrivateKey
+
+	for name, key := range cryptosshtestdata.PEMBytes {
+		err := os.Mkdir(filepath.Join(dir, name), os.ModePerm)
+		require.NoError(t, err)
+
+		filePath := filepath.Join(dir, name, name)
+		err = os.WriteFile(filePath, key, os.ModePerm)
+		require.NoError(t, err)
+
+		s, err := ssh.ParsePrivateKey(key)
+		require.NoError(t, err)
+
+		fingerprint := ssh.FingerprintSHA256(s.PublicKey())
+
+		key, err := accessgraph.NewPrivateKeyWithName(
+			privateKeyNameGen(filePath, deviceID, fingerprint),
+			&accessgraphsecretsv1pb.PrivateKeySpec{
+				PublicKeyFingerprint: fingerprint,
+				DeviceId:             deviceID,
+				PublicKeyMode:        accessgraphsecretsv1pb.PublicKeyMode_PUBLIC_KEY_MODE_DERIVED,
+			},
+		)
+		require.NoError(t, err)
+
+		expectedKeys = append(expectedKeys, key)
+	}
+
+	return expectedKeys
+}
+
+func writeEncryptedKeyWithoutPubFile(t *testing.T, dir string) []*accessgraphsecretsv1pb.PrivateKey {
```

Comment:
Doesn't this naturally happen on writeEncryptedKeys, by virtual of not all keys having "IncludesPublicKey: true"?

Reply:
What I meant is that this is already covered by test keys that do not have "IncludesPublicKey=true", right?